File: client/js/entity.js
define(['lib/underscore.min'], function () {

  var Entity = Class.extend({
    data: {
      name: "unknown",
      x: 0,
      y: 0
    },

    init: function (id, kind) {
      var self = this;

      this.interactable = true;

      this._name = "unknown";
      this._x = 0;
      this._y = 0;

      // Renderer
      this.nameOffsetY = -10;

      this.gridX = 0;
      this.gridY = 0;

      // Position
      this.setGridPosition(0, 0);


      if (id) {
        this.id = id;
      } else {
        this.id = Math.floor(Math.random() * 100000 + 100000);
      }

      this.kind = kind;

      // Renderer
      this.animations = {};

      // Modes
      this.isLoaded = false;

      this.reset();
    },

    reset: function () {
      // Renderer
      this.removed = false;
      this.flipSpriteX = false;
      this.flipSpriteY = false;
      this.currentAnimation = null;
      this.shadowOffsetY = 0;

      // Modes
      this.isHighlighted = false;
      this.visible = true;
      this.isFading = false;
      this.dirty();
    },

    get sprite() {
      return this._sprite;
    },

    set sprite(sprite) {
      this._sprite = sprite;
    },

    get x() {
      return this._x;
    },

    get y() {
      return this._y;
    },

    set x(x) {
      this._x = x;
    },

    set y(y) {
      this._y = y;
    },

    get name() {
      return this._name;
    },

    set name(name) {
      this._name = name;
    },

    setPosition: function (x, y) {
      this.x = x;
      this.y = y;

      this.trigger("PositionChange");
    },

    setGridPosition: function (x, y) {
      this.gridX = x;
      this.gridY = y;

      this.setPosition(x * 16, y * 16);
    },

    moveSteps: function (steps, orientation) {
      var gridX = this.gridX,
        gridY = this.gridY;

      switch (orientation) {
      case Types.Orientations.LEFT:
        gridX -= steps;
        break;
      case Types.Orientations.UP:
        gridY -= steps;
        break;
      case Types.Orientations.RIGHT:
        gridX += steps;
        break;
      case Types.Orientations.DOWN:
        gridY += steps;
        break;
      }

      this.setGridPosition(gridX, gridY);
    },


    setSprite: function (sprite, kindString) {
      if (!sprite) {
        throw "Error: " + this.id + " sprite is null (" + kindString + ")";
      }

      if (this.sprite && this.sprite.name === sprite.name) {
        return;
      }

      this.sprite = sprite;
      this.normalSprite = this.sprite;

      if (Types.isMob(this.kind) || Types.isPlayer(this.kind)) {
        this.hurtSprite = sprite.getHurtSprite();
      }

      _.extend(this.animations, sprite.createAnimations());

      this.isLoaded = true;
      if (this.ready_func) {
        this.ready_func();
      }
    },

    get skin() {
      return this.kind;
    },

    getSprite: function () {
      return this.sprite;
    },

    getSpriteName: function ()  {
      return Types.getKindAsString(this.kind);
    },

    get kindName() {
      return Types.getKindAsString(this.kind);
    },

    getAnimationByName: function (name) {
      var animations = this.animations;
      var animation = null;

      if (name in animations) {
        animation = animations[name];
      } else {
        log.error("No animation called " + name);
      }
      return animation;
    },

    setAnimation: function (name, speed, count, onEndCount) {
      var self = this;

      if (this.isLoaded) {
        if (this.currentAnimation && this.currentAnimation.name === name) {
          return;
        }

        var s = this.sprite,
          a = this.getAnimationByName(name);

        if (a) {
          this.currentAnimation = a;
          if (name.substr(0, 3) === "atk") {
            this.currentAnimation.reset();
          }
          this.currentAnimation.setSpeed(speed);
          this.currentAnimation.setCount(count ? count : 0, onEndCount || function () {
            self.idle();
          });
        }
      } else {
        this.log_error("Not ready for animation");
      }
    },

    hasShadow: function () {
      return false;
    },

    ready: function (f) {
      this.ready_func = f;
    },

    clean: function () {
      this.stopBlinking();
    },

    log_info: function (message) {
      log.info("[" + this.id + "] " + message);
    },

    log_error: function (message) {
      log.error("[" + this.id + "] " + message);
    },

    setHighlight: function (value) {
      if (value === true) {
        this.sprite = this.sprite.silhouetteSprite;
        this.isHighlighted = true;
      } else {
        this.sprite = this.normalSprite;
        this.isHighlighted = false;
      }
    },

    setVisible: function (value) {
      this.visible = value;
    },

    isVisible: function () {
      return this.visible;
    },

    toggleVisibility: function () {
      if (this.visible) {
        this.setVisible(false);
      } else {
        this.setVisible(true);
      }
    },

    isHostile: function (entity) {
      return false;
    },

    distanceTo: function (entity) {
      var distance = Math.sqrt(
        Math.pow(this.gridX - entity.gridX, 2) + Math.pow(this.gridY - entity.gridY, 2)
      );
      // round to 3 digit precision
      return Math.round(distance * Math.pow(10, 3)) / Math.pow(10, 3);
    },

    /**
     *
     */
    getDistanceToEntity: function (entity) {
      var distX = Math.abs(entity.gridX - this.gridX);
      var distY = Math.abs(entity.gridY - this.gridY);

      return (distX > distY) ? distX : distY;
    },

    isCloseTo: function (entity) {
      var dx, dy, d, close = false;
      if (entity) {
        dx = Math.abs(entity.gridX - this.gridX);
        dy = Math.abs(entity.gridY - this.gridY);

        if (dx < 30 && dy < 14) {
          close = true;
        }
      }
      return close;
    },

    /**
     * Returns true if the entity is adjacent to the given one.
     * @returns {Boolean} Whether these two entities are adjacent.
     */
    isAdjacent: function (entity) {
      var adjacent = false;

      if (entity) {
        adjacent = this.getDistanceToEntity(entity) > 1 ? false : true;
      }
      return adjacent;
    },

    /**
     *
     */
    isAdjacentNonDiagonal: function (entity) {
      var result = false;

      if (this.isAdjacent(entity) && !(this.gridX !== entity.gridX && this.gridY !== entity.gridY)) {
        result = true;
      }

      return result;
    },

    isDiagonallyAdjacent: function (entity) {
      return this.isAdjacent(entity) && !this.isAdjacentNonDiagonal(entity);
    },

    forEachAdjacentNonDiagonalPosition: function (callback) {
      callback(this.gridX - 1, this.gridY, Types.Orientations.LEFT);
      callback(this.gridX, this.gridY - 1, Types.Orientations.UP);
      callback(this.gridX + 1, this.gridY, Types.Orientations.RIGHT);
      callback(this.gridX, this.gridY + 1, Types.Orientations.DOWN);

    },

    fadeIn: function (currentTime) {
      this.isFading = true;
      this.startFadingTime = currentTime;
    },

    blink: function (speed, callback) {
      var self = this;

      this.blinking = setInterval(function () {
        self.toggleVisibility();
      }, speed);
    },

    stopBlinking: function () {
      if (this.blinking) {
        clearInterval(this.blinking);
      }
      this.setVisible(true);
    },

    dirty: function () {
      this.isDirty = true;
      this.trigger("dirty");
    }
  });

  return Entity;
});

File: client/dart/entity.dart
library entity;

import "dart:async";
import "dart:html" as html;
import "dart:math";

import "animation.dart";
import "base.dart";
import "game.dart";
import "position.dart";
import "rect.dart";
import "sprite.dart";
import "lib/gametypes.dart";

class Entity extends Base {
  int id;
  EntityKind kind;
  bool interactable = true;
  String name;
  
  int x = 0;
  int y = 0;
  Position _gridPosition = const Position(0, 0);
  Orientation orientation = Orientation.DOWN;

  bool isDead = false;
  bool isDying = false;
  bool isLoaded = false;
  bool isHighlighted = false;
  bool isVisible = true;
  bool isDirty = false;
  bool isOnPlateau = false;

  bool isFading = false;
  num fadingAlpha = 1;
  int startFadingTime;

  Timer blinkingTimer;

  // Renderer
  Sprite _sprite;
  Map<String, Animation> animations = {};
  Animation currentAnimation;
  int nameOffsetY = -10;
  int shadowOffsetY = 0;
  bool isRemoved = false;
  bool flipSpriteX = false; 
  bool flipSpriteY = false; 
  Rect dirtyRect;
  Rect oldDirtyRect;

  Entity(int this.id, EntityKind this.kind);

  Position get gridPosition => this._gridPosition;
  void set gridPosition(Position position) {
    if (position == null) {
      throw new Exception("position set to null!");
    }
    this._gridPosition = position;
    this.x = position.x * 16;
    this.y = position.y * 16;
    this.trigger("PositionChange");
  }

  Sprite get sprite {
    if (this._sprite == null) {
      return null;
    }

    return this.isHighlighted ? this._sprite.getSilhouetteSprite() : this._sprite;
  }

  void set sprite(Sprite sprite) {
    this._sprite = sprite;
  }

  void reset() {
    this.orientation = Orientation.DOWN;

    // Renderer
    this.isRemoved = false;
    this.flipSpriteX = false;
    this.flipSpriteY = false;
    this.currentAnimation = null;
    this.shadowOffsetY = 0;

    // Modes
    this.isDead = false;
    this.isDying = false;
    this.isHighlighted = false;
    this.isVisible = true;
    this.isFading = false;
    this.dirty();
  }
  
  void die() {
    html.window.console.info("${this.id} is dead");

    this.isDead = true;
    this.isDying = true;
    this.setSprite(Game.sprites["death"]);

    this.animate("death", 120, 1, () {
      html.window.console.info("${this.id} was removed");

      Game.removeEntity(this);
      Game.removeFromRenderingGrid(this, this.gridPosition);
    });

    // Upon death, this entity is removed from both grids, allowing the player
    // to click very fast in order to loot the dropped item and not be blocked.
    // The entity is completely removed only after the death animation has ended.
    Game.removeFromEntityGrid(this, this.gridPosition);
    Game.removeFromPathingGrid(this.gridPosition);

    Game.updateCursor();
  }

  /**
   * Returns whether we changed to the new animation or not.
   */
  bool animate(String animationName, int speed, [int count = 0, Function onEndCount]) {
    // don't change animation if the character is dying
    if (this.currentAnimation != null 
        && this.currentAnimation.name == "death" 
        && this.isDying) {
      return false;
    }

    this.flipSpriteX = false;
    this.flipSpriteY = false;
    List<String> oriented = ['atk', 'walk', 'idle'];
    if (oriented.contains(animationName)) {
      animationName += "_";
      animationName += 
        this.orientation == Orientation.LEFT 
        ? Types.getOrientationAsString(Orientation.RIGHT)
        : Types.getOrientationAsString(this.orientation);
      this.flipSpriteX = this.orientation == Orientation.LEFT;
    }

    this.setAnimation(animationName, speed, count, onEndCount);
    return true;
  }

  void moveSteps(int steps, Orientation orientation) {
    int gridX = this.gridPosition.x;
    int gridY = this.gridPosition.y;

    switch (orientation) {
      case Orientation.LEFT:
        gridX -= steps;
        break;
      case Orientation.UP:
        gridY -= steps;
        break;
      case Orientation.RIGHT:
        gridX += steps;
        break;
      case Orientation.DOWN:
        gridY += steps;
        break;
    }

    this.gridPosition = new Position(gridX, gridY);
  }

  void setSprite(Sprite sprite) {
    if (sprite == null) {
      throw "cannot setSprite to null for entity ${this.id}";
    }

    // don't change to the same sprite
    if (this._sprite != null && this._sprite.name == sprite.name) {
      return;
    }

    this._sprite = sprite;
    this.animations.addAll(sprite.createAnimations());

    this.isLoaded = true;
    this.trigger("Ready");
  }

  EntityKind get skin => this.kind;

  String getSpriteName() => Types.getKindAsString(this.kind);

  void idle([Orientation orientation]) {}

  void setAnimation(String name, int speed, [int count = 0, Function onEndCount]) {
    if (!this.isLoaded) {
      this.log_error("Not ready for animation");
    }

    // if we are already animating the given animation, stop.
    if (this.currentAnimation != null && this.currentAnimation.name == name) {
      return;
    }

    Animation animation = this.animations[name];
    this.currentAnimation = animation;
    if (name.substring(0, 3) == "atk") {
      this.currentAnimation.reset();
    }

    this.currentAnimation.speed = speed;
    this.currentAnimation.count = count;
    this.currentAnimation.on(
      "EndCount", 
      onEndCount is Function ? onEndCount : () { this.idle(); }, 
      /* overwrite old callbacks */ true
    );
  }

  bool hasShadow() => false;

  void clean() {
    this.stopBlinking();
  }

  void log_debug(String message) {
    html.window.console.debug("[$this.id] $message");
  }

  void log_info(String message) {
    html.window.console.info("[$this.id] $message");
  }

  void log_error(String message) {
    html.window.console.error("[$this.id] $message");
  }

  bool isHostile(Entity entity) => false;

  num distanceTo(Entity entity) => sqrt(pow(this.gridPosition.x - entity.gridPosition.x, 2) + pow(this.gridPosition.y - entity.gridPosition.y, 2));

  int getDistanceToEntity(Entity entity) => max((this.gridPosition.x - entity.gridPosition.x).abs(), (this.gridPosition.y - entity.gridPosition.y).abs());

  bool isCloseTo(Entity entity) => (this.gridPosition.x - entity.gridPosition.x).abs() < 30 && (this.gridPosition.y - entity.gridPosition.y).abs() < 14;

  bool isAdjacent(Entity entity) => this.getDistanceToEntity(entity) > 1 ? false : true;

  bool isAdjacentNonDiagonal(Entity entity) => this.isAdjacent(entity) && !(this.gridPosition.x != entity.gridPosition.x && this.gridPosition.y != entity.gridPosition.y);

  bool isDiagonallyAdjacent(Entity entity) => this.isAdjacent(entity) && !this.isAdjacentNonDiagonal(entity);

  void forEachAdjacentNonDiagonalPosition(callback) {
    callback(this.gridPosition.x - 1, this.gridPosition.y, Orientation.LEFT);
    callback(this.gridPosition.x, this.gridPosition.y - 1, Orientation.UP);
    callback(this.gridPosition.x + 1, this.gridPosition.y, Orientation.RIGHT);
    callback(this.gridPosition.x, this.gridPosition.y + 1, Orientation.DOWN);
  }

  void fadeIn(int currentTime) {
    this.isFading = true;
    this.startFadingTime = currentTime;
  }

  void blink(int speed, Function callback) {
    this.blinkingTimer = new Timer.periodic(
      new Duration(milliseconds: speed),
      (Timer timer) { this.isVisible = !this.isVisible; }
    );
  }

  void stopBlinking() {
    if (this.blinkingTimer != null && this.blinkingTimer.isActive) {
      this.blinkingTimer.cancel();
    }

    this.isVisible = true;
  }

  void dirty() {
    this.isDirty = true;
    this.trigger("Dirty");
  }
}
